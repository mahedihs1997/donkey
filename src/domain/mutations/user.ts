import gql from 'graphql-tag';

export const CREATE_USER_QUERY = gql`
  mutation CreateUser($email: String!, $password: String!) {
    loginUser(data: { email: $email, password: $password }) {
      token
    }
  }
`;
