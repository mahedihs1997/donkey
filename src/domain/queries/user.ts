import gql from 'graphql-tag';

export const ME_QUERY = gql`
  query Me {
    me {
      username
      role
    }
  }
`;

export const USERS_LIST_QUERY = gql`
  query Users {
    users {
      username
      name
    }
  }
`;
