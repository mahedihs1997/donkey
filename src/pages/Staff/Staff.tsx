import { useQuery } from '@apollo/react-hooks';
import CreateUser, { CreateUserRefs } from '@components/CreateUser';
import VirtualizedTable from '@components/VirtualizedTable';
import { USERS_LIST_QUERY } from '@domain/queries/user';
import { Button, IconButton, Paper } from '@material-ui/core';
import Divider from '@material-ui/core/Divider';
import Typography from '@material-ui/core/Typography';
import { Edit } from '@material-ui/icons';
import { NextPage } from 'next';
import React from 'react';

import { useStyles } from './Staff.styles';

const Staff: NextPage = () => {
  const { data, loading, refetch, networkStatus } = useQuery(USERS_LIST_QUERY);

  const modal = React.useRef<CreateUserRefs>(null);

  const handleOpen = () => {
    modal.current?.handleOpen();
  };

  const classes = useStyles();

  const rows = React.useMemo(() => {
    if (loading || !data) {
      return [];
    }

    return data.users.map((user) => ({
      ...user,
      role: 'Admin',
      rate: 'Â£1.75',
      edit: (
        <IconButton>
          <Edit color="secondary" />
        </IconButton>
      ),
    }));
  }, [loading, networkStatus]);

  const onCreateUser = () => {
    refetch();
  };

  return (
    <div>
      <div className={classes.header}>
        <Typography variant="h4" className={classes.title}>
          List of users
        </Typography>
        <Button
          type="button"
          onClick={handleOpen}
          variant="contained"
          color="primary"
        >
          Create User
        </Button>
      </div>
      <Divider className={classes.divider} />
      <Paper style={{ height: 432, maxWidth: 1000 }}>
        <VirtualizedTable
          rowCount={rows.length}
          rowGetter={({ index }) => rows[index]}
          columns={[
            {
              width: 200,
              label: 'Name',
              dataKey: 'name',
            },
            {
              width: 200,
              label: 'Username',
              dataKey: 'username',
            },
            {
              width: 200,
              label: 'Role',
              dataKey: 'role',
            },
            {
              width: 200,
              label: 'Rate',
              dataKey: 'rate',
            },
            {
              width: 200,
              label: 'Edit',
              dataKey: 'edit',
            },
          ]}
        />
      </Paper>
      <CreateUser onSuccess={onCreateUser} ref={modal} />
    </div>
  );
};

Staff.getInitialProps = async () => {
  return {
    showLayout: true,
    title: 'Staff',
  };
};

export default Staff;
