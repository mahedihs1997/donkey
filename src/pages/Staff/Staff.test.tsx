import { MockedProvider } from '@apollo/react-testing';
import { USERS_LIST_QUERY } from '@domain/queries/user';
import React from 'react';
import TestRenderer from 'react-test-renderer';

import Staff from '.';

const mocks = [
  {
    request: {
      query: USERS_LIST_QUERY,
    },
    result: {
      data: {
        users: [{ name: 'Ugendo', username: 'Bugendo' }],
      },
    },
  },
];

describe('<Staff />', () => {
  it('should render without throwing an error and match snapshot', () => {
    const testRenderer = TestRenderer.create(
      <MockedProvider addTypename={false} mocks={mocks}>
        <Staff />
      </MockedProvider>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});
