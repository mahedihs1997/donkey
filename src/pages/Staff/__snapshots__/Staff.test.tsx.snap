// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Staff /> should render without throwing an error and match snapshot 1`] = `
<div>
  <div
    className="makeStyles-header-3"
  >
    <h4
      className="MuiTypography-root makeStyles-title-4 MuiTypography-h4"
    >
      List of users
    </h4>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <span
        className="MuiButton-label"
      >
        Create User
      </span>
      <span
        className="MuiTouchRipple-root"
      />
    </button>
  </div>
  <hr
    className="MuiDivider-root makeStyles-divider-2"
  />
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
    style={
      Object {
        "height": 432,
        "maxWidth": 1000,
      }
    }
  >
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    >
      <div
        aria-colcount={5}
        aria-rowcount={0}
        className="ReactVirtualized__Table MuiVirtualizedTable-table-6"
        role="grid"
        style={Object {}}
      >
        <div
          className="ReactVirtualized__Table__headerRow MuiVirtualizedTable-tableRow-7 MuiVirtualizedTable-flexContainer-5"
          role="row"
          style={
            Object {
              "height": 48,
              "overflow": "hidden",
              "paddingRight": 0,
              "width": 0,
            }
          }
        >
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignLeft"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Name
              </span>
            </div>
          </div>
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignLeft"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Username
              </span>
            </div>
          </div>
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignLeft"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Role
              </span>
            </div>
          </div>
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignLeft"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Rate
              </span>
            </div>
          </div>
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignLeft"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Edit
              </span>
            </div>
          </div>
        </div>
        <div
          aria-label="grid"
          aria-readonly={null}
          className="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
          onScroll={[Function]}
          role="rowgroup"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "boxSizing": "border-box",
              "direction": "inherit",
              "height": -48,
              "overflowX": "hidden",
              "overflowY": "auto",
              "position": "relative",
              "width": 0,
              "willChange": "transform",
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
  </div>
</div>
`;
